### DISCUSSION TITLE
onAudioEnd implementation

### DISCUSSION THREAD

**Discussion by Jakob-SD** [mediagrid/capacitor-native-audio#13] [1. Help]
Hi there

First let me thank you for creating this awesome plugin! 👍

My implementation of AudioPlayer.onAudioEnd is failing. The event is never called when the audio file is finished playing. The other event listeners are working fine, and I can play, pause, stop and seek without issues. Do I need to use it in a different way?

Thanks 🙏

I get no errors from the app.

```javascript

async init() {

    const url = await this.StoryService.getStoryLink(this.story);
    this.storyLink = url;

    await AudioPlayer.create({
      audioId: this.story.token,
      audioSource: this.storyLink,
      friendlyTitle: this.story.title,
      useForNotification: true,
      artworkSource: '',
      isBackgroundMusic: false,
      loop: false,
    }).catch(ex => setError(ex));

    await AudioPlayer.onAudioReady({ audioId: this.story.token }, async () => {
      const duration = await getDuration(this.story.token);
      this.duration = duration;
    });

    await AudioPlayer.onAudioEnd({ audioId: this.story.token }, async () => {
      this.stopCurrentPositionUpdate();
      console.log('Audio ended');
      this.stop();
      this.currentTime = 0;
    });

    await AudioPlayer.onAppGainsFocus({ audioId: this.story.token }, async () => {
      console.log('I have focus');
    });

    await AudioPlayer.onAppLosesFocus({ audioId: this.story.token }, () => {
      console.log('I lost focus');
    });

    await AudioPlayer.initialize({ audioId: this.story.token }).catch(ex => console.log(ex));
  }
```

**Comment by wsamoht** [comment]
Nothing looks wrong in your code. What device/emulator are you testing on? Did you check XCode logs and/or LogCat in Android Studio as well?

**Reply by Jakob-SD** [reply]
Hi, 

I'm testing with different devices through xcode - both virtual and real devices. I haven't moved on to Android yet. I use xcode log system to debug. Currently I'm just comparing the total length of the audio with the timer value, and use that to handle my logic when audio ends. Seems to work fine, so the onAudioEnd listener is not necessary (for now at least)

**Reply by wsamoht** [reply]
Nice workaround. I am not sure why it wouldn't be working 😬 What versions of iOS are you testing?

**Reply by Jakob-SD** [reply]
iOS 18.0.1 on my phones and and iOS 18.0 in simulator.

**Reply by wsamoht** [reply]
Got it. I don't think I've tested the plugin on iOS 18 yet. I don't own any Apple devices (we use a rented Mac in the cloud for our app builds, but I cannot hear audio in the emulator since I am remoted in). Gotta love Apple's closed ecosystem.... I am asking my business partner to see if he has upgraded to iOS 18 and whether he is seeing issues in our app.

**Reply by wsamoht** [reply]
An update from my end. My business partner checked our app on iOS 18 and he didn't seem to have any issues 🤔
